examples/single_mem_node/main.rs

```diff
@@ -0,0 +1,202 @@
+        // it doesn't receive any message from the leader.
+        election_tick: 10,
+        // Heartbeat tick is for how long the leader needs to send
+        // a heartbeat to keepalive.
+        heartbeat_tick: 3,
+        // The max size limits the max size of each append message, mostly, 1 MB is enough.
+        max_size_per_msg: 1024 * 1024 * 1024,
+        // Max inflight msgs that the leader sends message to follower without
+        // receiving ACKs.
+        max_inflight_msgs: 256,
+        // The Raft applied index.
+        // You need to save your applied index when you apply the committed Raft logs.
+        applied: 0,
+        // Just for log
+        tag: format!("[{}]", 1),
+        ..Default::default()
+    };
+
+    // Create the Raft node.
+    let mut r = RawNode::new(&cfg, storage, vec![]).unwrap();
+
+    let (sender, receiver) = mpsc::channel();
+
+    // Use another thread to propose a Raft request.
+    send_propose(sender);
+
+    // Loop forever to drive the Raft.
+    let mut t = Instant::now();
+    let mut timeout = Duration::from_millis(100);
+
+    // Use a HashMap to holde the propose callbacks.
+    let mut cbs = HashMap::new();
+
+    loop {
+        match receiver.recv_timeout(timeout) {
+            Ok(Msg::Propose { id, cb }) => {
+                cbs.insert(id, cb);
+                r.propose(vec![id], false).unwrap();
+            }
+            Ok(Msg::Raft(m)) => r.step(m).unwrap(),
+            Err(RecvTimeoutError::Timeout) => (),
+            Err(RecvTimeoutError::Disconnected) => return,
+        }
+
+        let d = t.elapsed();
+        if d >= timeout {
+            t = Instant::now();
+            timeout = Duration::from_millis(100);
+            // We drive Raft every 100ms.
+            r.tick();
+        } else {
+            timeout -= d;
+        }
+
+        on_ready(&mut r, &mut cbs);
+    }
+}
+
+fn on_ready(r: &mut RawNode<MemStorage>, cbs: &mut HashMap<u8, ProposeCallback>) {
+    if !r.has_ready() {
+        return;
+    }
+
+    // The Raft is ready, we can do something now.
+    let mut ready = r.ready();
+
+    let is_leader = r.raft.leader_id == 1;
+    if is_leader {
+        // If the peer is leader, the leader can send messages to other followers ASAP.
+        let msgs = ready.messages.drain(..);
+        for _msg in msgs {
+            // Here we only have one peer, so can ignore this.
+            // Mostly, for optimization, we can send messages
+        }
+    }
+
+    if !raft::is_empty_snap(&ready.snapshot) {
+        // This is a snapshot, we need to apply the snapshot at first.
+        r.mut_store()
+            .wl()
+            .apply_snapshot(ready.snapshot.clone())
+            .unwrap();
+    }
+
+    if !ready.entries.is_empty() {
+        // Append entries to the Raft log
+        r.mut_store().wl().append(&ready.entries).unwrap();
+    }
+
+    if let Some(ref hs) = ready.hs {
+        // Raft HardState changed, we need to persist it.
```

Comment:
-> Raft HardState changed. we need to persist it.
OR
-> Raft HardState changed, and we need to persist it.